File: python/orderbookStreamExample/orderbook_stream_example.py
#!/usr/bin/env python3
"""
Orderbook Stream Example - Stream Hyperliquid orderbook data via QuickNode gRPC.

Setup:
    pip install -r requirements.txt
    python -m grpc_tools.protoc -I../../proto --python_out=. --grpc_python_out=. ../../proto/orderbook.proto

Required environment:
    export GRPC_ENDPOINT="your-endpoint.hype-mainnet.quiknode.pro:10000"
    export AUTH_TOKEN="YOUR_QUICKNODE_TOKEN"
"""

import argparse
import json
import os
import sys
import time
from typing import Callable, Iterable, Optional

import grpc

try:
    import orderbook_pb2 as pb
    import orderbook_pb2_grpc as pb_grpc
except ImportError:
    print("Error: Proto files not generated. Run:")
    print("  python -m grpc_tools.protoc -I../../proto --python_out=. --grpc_python_out=. ../../proto/orderbook.proto")
    sys.exit(1)

GRPC_ENDPOINT = os.environ.get("GRPC_ENDPOINT", "your-endpoint.hype-mainnet.quiknode.pro:10000")
AUTH_TOKEN = os.environ.get("AUTH_TOKEN") or os.environ.get("QN_AUTH_TOKEN") or "your-quicknode-token"
MAX_RETRIES = 10
BASE_DELAY_SECONDS = 2


def create_channel() -> grpc.Channel:
    return grpc.secure_channel(
        GRPC_ENDPOINT,
        grpc.ssl_channel_credentials(),
        options=[
            ("grpc.max_receive_message_length", 100 * 1024 * 1024),
            ("grpc.keepalive_time_ms", 30000),
        ],
    )


def metadata():
    return [("x-token", AUTH_TOKEN)]


def level_text(level) -> str:
    if not level or not level.px:
        return "n/a"
    return f"{level.px} / {level.sz} ({level.n})"


def enum_name(enum_type, value: int) -> str:
    try:
        return enum_type.Name(value)
    except ValueError:
        return str(value)


def l4_snapshot_reset_kind(snapshot_count: int) -> str:
    if isinstance(snapshot_count, bool) or not isinstance(snapshot_count, int) or snapshot_count < 1:
        raise ValueError("snapshot_count must be a positive integer")
    return "initial" if snapshot_count == 1 else "replacement"


def l2_request(args) -> pb.L2BookRequest:
    request = pb.L2BookRequest(coin=args.coin, n_levels=args.levels)
    if args.sig_figs is not None:
        request.n_sig_figs = args.sig_figs
    if args.mantissa is not None:
        request.mantissa = args.mantissa
    return request


def l2_diff_request(args) -> pb.L2BookDiffRequest:
    request = pb.L2BookDiffRequest(
        coins=args.coins,
        n_levels=args.levels,
        skip_initial_snapshot=args.skip_initial_snapshot,
    )
    if args.sig_figs is not None:
        request.n_sig_figs = args.sig_figs
    if args.mantissa is not None:
        request.mantissa = args.mantissa
    return request


def consume_with_reconnect(
    label: str,
    make_stream: Callable[[pb_grpc.OrderBookStreamingStub], Iterable],
    handle_update: Callable[[object, int], None],
    max_messages: Optional[int],
):
    msg_count = 0
    data_loss_count = 0

    while data_loss_count < MAX_RETRIES:
        if data_loss_count > 0:
            delay = BASE_DELAY_SECONDS * (2 ** (data_loss_count - 1))
            print(f"Reconnecting {label} after DATA_LOSS in {delay}s (attempt {data_loss_count + 1}/{MAX_RETRIES})")
            time.sleep(delay)

        channel = create_channel()
        stub = pb_grpc.OrderBookStreamingStub(channel)

        try:
            for update in make_stream(stub):
                data_loss_count = 0
                msg_count += 1
                handle_update(update, msg_count)

                if max_messages and msg_count >= max_messages:
                    return

            return
        except grpc.RpcError as exc:
            if exc.code() == grpc.StatusCode.DATA_LOSS:
                data_loss_count += 1
                channel.close()
                continue
            raise
        finally:
            channel.close()

    raise RuntimeError(f"{label} exceeded max reconnect attempts")


def stream_l2(args):
    consume_with_reconnect(
        "StreamL2Book",
        lambda stub: stub.StreamL2Book(l2_request(args), metadata=metadata()),
        lambda update, count: print(
            f"[{count}] L2 {update.coin} block={update.block_number} "
            f"bid={level_text(update.bids[0] if update.bids else None)} "
            f"ask={level_text(update.asks[0] if update.asks else None)} "
            f"bids={len(update.bids)} asks={len(update.asks)}"
        ),
        args.max_messages,
    )


def stream_l4(args):
    snapshot_count = 0

    def handle(update, count):
        nonlocal snapshot_count
        if update.HasField("snapshot"):
            snapshot_count += 1
            snapshot = update.snapshot
            reset = l4_snapshot_reset_kind(snapshot_count)
            print(
                f"[{count}] L4 snapshot {snapshot.coin} height={snapshot.height} "
                f"reset={reset} bids={len(snapshot.bids)} asks={len(snapshot.asks)}"
            )
            if reset == "replacement":
                print("  replace the entire local L4 book with this snapshot")
        elif update.HasField("diff"):
            try:
                data = json.loads(update.diff.data)
            except json.JSONDecodeError as exc:
                print(f"[{count}] L4 diff height={update.diff.height} invalid JSON: {exc}")
                return
            print(
                f"[{count}] L4 diff height={update.diff.height} "
                f"order_statuses={len(data.get('order_statuses', []))} "
                f"book_diffs={len(data.get('book_diffs', []))}"
            )

    consume_with_reconnect(
        "StreamL4Book",
        lambda stub: stub.StreamL4Book(pb.L4BookRequest(coin=args.coin), metadata=metadata()),
        handle,
        args.max_messages,
    )


def stream_bbo(args):
    consume_with_reconnect(
        "StreamBboBook",
        lambda stub: stub.StreamBboBook(pb.BboBookRequest(coins=args.coins), metadata=metadata()),
        lambda update, count: print(
            f"[{count}] BBO {update.coin} block={update.block_number} "
            f"bid={level_text(update.bid if update.HasField('bid') else None)} "
            f"ask={level_text(update.ask if update.HasField('ask') else None)}"
        ),
        args.max_messages,
    )


def stream_l2_diff(args):
    def handle(update, count):
        print(f"[{count}] L2 diff height={update.height} snapshot={update.snapshot} coins={len(update.diffs)}")
        for diff in list(update.diffs)[:5]:
            print(
                f"  {diff.coin} seq={diff.seq} prev_seq={diff.prev_seq} "
                f"snapshot={diff.snapshot} bid_changes={len(diff.bids)} ask_changes={len(diff.asks)}"
            )

    consume_with_reconnect(
        "StreamL2BookDiff",
        lambda stub: stub.StreamL2BookDiff(l2_diff_request(args), metadata=metadata()),
        handle,
        args.max_messages,
    )


def stream_l4_updates(args):
    def handle(update, count):
        print(f"[{count}] L4 updates height={update.height} snapshot={update.snapshot} diffs={len(update.diffs)}")
        if update.snapshot:
            print("  clear local L4 order state before applying this update")
        for diff in list(update.diffs)[:5]:
            print(
                f"  {enum_name(pb.L4OrderDiffType, diff.diff_type)} {diff.coin} "
                f"oid={diff.oid} side={diff.side or 'n/a'} px={diff.px or 'n/a'} sz={diff.sz or 'n/a'}"
            )

    consume_with_reconnect(
        "StreamL4BookUpdates",
        lambda stub: stub.StreamL4BookUpdates(pb.L4BookUpdatesRequest(coins=args.coins), metadata=metadata()),
        handle,
        args.max_messages,
    )


def stream_tpsl(args):
    def handle(update, count):
        print(f"[{count}] TP/SL height={update.height} snapshot={update.snapshot} diffs={len(update.diffs)}")
        for diff in list(update.diffs)[:5]:
            print(
                f"  {enum_name(pb.TpslDiffType, diff.diff_type)} {diff.coin} "
                f"oid={diff.oid} trigger={diff.trigger_px or 'n/a'} "
                f"limit={diff.limit_px or 'n/a'} sz={diff.sz or 'n/a'} reason={diff.reason or 'n/a'}"
            )

    consume_with_reconnect(
        "StreamTpslUpdates",
        lambda stub: stub.StreamTpslUpdates(pb.TpslUpdatesRequest(coins=args.coins), metadata=metadata()),
        handle,
        args.max_messages,
    )


def parse_args():
    parser = argparse.ArgumentParser(description="Stream Hyperliquid orderbook data via QuickNode gRPC")
    parser.add_argument("--mode", choices=["l2", "l4", "bbo", "l2-diff", "l4-updates", "tpsl"], default="bbo")
    parser.add_argument("--coin", default="BTC", help="Coin symbol or comma-separated symbols")
    parser.add_argument("--all", action="store_true", help="Subscribe to all eligible coins on multi-coin streams")
    parser.add_argument("--levels", type=int, default=20, help="Number of L2 levels (default 20, max 100)")
    parser.add_argument("--sig-figs", type=int, default=None, help="L2 bucketing significant figures (2-5)")
    parser.add_argument("--mantissa", type=int, default=None, help="L2 bucketing mantissa (1, 2, or 5)")
    parser.add_argument("--skip-initial-snapshot", action="store_true", help="For l2-diff, skip the initial snapshot")
    parser.add_argument("--max-messages", type=int, default=None, help="Stop after N messages")
    args = parser.parse_args()

    if args.all and args.mode in {"l2", "l4"}:
        parser.error("--all is only supported for bbo, l2-diff, l4-updates, and tpsl. Use --coin for l2 or l4.")

    args.coins = [] if args.all else [coin.strip() for coin in args.coin.split(",") if coin.strip()]
    if not args.all and not args.coins:
        parser.error("--coin must include at least one symbol. Use --all to subscribe to every eligible coin on multi-coin streams.")

    args.coin = args.coins[0] if args.coins else ""
    return args


def coin_display(args) -> tuple[str, str]:
    if args.mode in {"l2", "l4"}:
        return "Coin", args.coin
    return "Coins", ",".join(args.coins) if args.coins else "all eligible coins"


def main():
    args = parse_args()

    print("Hyperliquid Orderbook Stream Example")
    print(f"Endpoint: {GRPC_ENDPOINT}")
    print(f"Mode: {args.mode}")
    coin_label, coin_value = coin_display(args)
    print(f"{coin_label}: {coin_value}")

    if AUTH_TOKEN == "your-quicknode-token":
        print("Set AUTH_TOKEN to your QuickNode token before running this example.")
        sys.exit(1)

    if args.mode == "l2":
        stream_l2(args)
    elif args.mode == "l4":
        stream_l4(args)
    elif args.mode == "bbo":
        stream_bbo(args)
    elif args.mode == "l2-diff":
        stream_l2_diff(args)
    elif args.mode == "l4-updates":
        stream_l4_updates(args)
    elif args.mode == "tpsl":
        stream_tpsl(args)


if __name__ == "__main__":
    main()

File: javascript/orderbookStreamExample/orderbook_stream_example.js
// Orderbook Stream Example - Stream Hyperliquid orderbook data via QuickNode gRPC
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

// Mainnet: your-endpoint.hype-mainnet.quiknode.pro:10000
// Testnet: your-endpoint.hype-testnet.quiknode.pro:10000
const GRPC_ENDPOINT = process.env.GRPC_ENDPOINT || 'your-endpoint.hype-mainnet.quiknode.pro:10000';
const AUTH_TOKEN = process.env.AUTH_TOKEN || process.env.QN_AUTH_TOKEN || 'your-quicknode-token';
const GRPC_PLAINTEXT = process.env.GRPC_PLAINTEXT === '1';
const PROTO_PATH = path.join(__dirname, '..', '..', 'proto', 'orderbook.proto');
const MAX_RECEIVE_BYTES = 100 * 1024 * 1024;
const L4_FLOW_CONTROL_BYTES = 32 * 1024 * 1024;

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});
const proto = grpc.loadPackageDefinition(packageDefinition).hyperliquid;

function channelOptions() {
  return {
    'grpc.max_receive_message_length': MAX_RECEIVE_BYTES,
    // BTC L4 snapshots contain tens of thousands of orders. A larger HTTP/2
    // receive window prevents the server from timing out while Node decodes them.
    'grpc-node.flow_control_window': L4_FLOW_CONTROL_BYTES,
    'grpc.keepalive_time_ms': 30000,
    'grpc.keepalive_timeout_ms': 10000
  };
}

function createClient() {
  return new proto.OrderBookStreaming(
    GRPC_ENDPOINT,
    GRPC_PLAINTEXT ? grpc.credentials.createInsecure() : grpc.credentials.createSsl(),
    channelOptions()
  );
}

function createMetadata() {
  const metadata = new grpc.Metadata();
  metadata.add('x-token', AUTH_TOKEN);
  return metadata;
}

function parseArgs() {
  const args = process.argv.slice(2);
  const get = (name, fallback = null) => {
    const arg = args.find(a => a.startsWith(`--${name}=`));
    return arg ? arg.split('=').slice(1).join('=') : fallback;
  };

  const coinArg = get('coin', 'BTC');
  const all = args.includes('--all');
  const coins = all ? [] : coinArg.split(',').map(c => c.trim()).filter(Boolean);
  return {
    mode: get('mode', 'bbo'),
    all,
    coins,
    coin: coins[0] || '',
    levels: parseInt(get('levels', '20'), 10),
    sigFigs: get('sig-figs') ? parseInt(get('sig-figs'), 10) : null,
    mantissa: get('mantissa') ? parseInt(get('mantissa'), 10) : null,
    skipInitialSnapshot: args.includes('--skip-initial-snapshot'),
    maxMessages: get('max-messages') ? parseInt(get('max-messages'), 10) : null
  };
}

function validateArgs(args) {
  if (args.all && (args.mode === 'l2' || args.mode === 'l4')) {
    console.error('--all is only supported for bbo, l2-diff, l4-updates, and tpsl. Use --coin for l2 or l4.');
    process.exit(2);
  }
  if (!args.all && args.coins.length === 0) {
    console.error('--coin must include at least one symbol. Use --all to subscribe to every eligible coin on multi-coin streams.');
    process.exit(2);
  }
}

function coinDisplay(args) {
  if (args.mode === 'l2' || args.mode === 'l4') {
    return { label: 'Coin', value: args.coin };
  }
  return {
    label: 'Coins',
    value: args.coins.length ? args.coins.join(',') : 'all eligible coins'
  };
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function levelText(level) {
  if (!level || !level.px) return 'n/a';
  return `${level.px} / ${level.sz} (${level.n})`;
}

function l4SnapshotResetKind(snapshotCount) {
  if (!Number.isInteger(snapshotCount) || snapshotCount < 1) {
    throw new Error('snapshotCount must be a positive integer');
  }
  return snapshotCount === 1 ? 'initial' : 'replacement';
}

function l2Request(args) {
  const request = {
    coin: args.coin,
    n_levels: args.levels
  };
  if (args.sigFigs !== null) request.n_sig_figs = args.sigFigs;
  if (args.mantissa !== null) request.mantissa = args.mantissa;
  return request;
}

function l2DiffRequest(args) {
  const request = {
    coins: args.coins,
    n_levels: args.levels,
    skip_initial_snapshot: args.skipInitialSnapshot
  };
  if (args.sigFigs !== null) request.n_sig_figs = args.sigFigs;
  if (args.mantissa !== null) request.mantissa = args.mantissa;
  return request;
}

async function consumeWithReconnect(label, makeCall, onData, maxMessages = null) {
  const maxRetries = 10;
  const baseDelayMs = 2000;
  let msgCount = 0;
  let dataLossCount = 0;

  while (dataLossCount < maxRetries) {
    if (dataLossCount > 0) {
      const delay = baseDelayMs * Math.pow(2, dataLossCount - 1);
      console.log(`Reconnecting ${label} after DATA_LOSS in ${delay / 1000}s (attempt ${dataLossCount + 1}/${maxRetries})`);
      await sleep(delay);
    }

    const result = await new Promise((resolve, reject) => {
      const call = makeCall();

      call.on('data', (update) => {
        dataLossCount = 0;
        msgCount += 1;
        onData(update, msgCount);

        if (maxMessages && msgCount >= maxMessages) {
          call.cancel();
        }
      });

      call.on('error', (err) => {
        if (err.code === grpc.status.CANCELLED && maxMessages && msgCount >= maxMessages) {
          resolve('done');
        } else if (err.code === grpc.status.DATA_LOSS) {
          resolve('data_loss');
        } else {
          reject(err);
        }
      });

      call.on('end', () => resolve('done'));
    });

    if (result === 'done') return;
    dataLossCount += 1;
  }

  throw new Error(`${label} exceeded max reconnect attempts`);
}

async function streamL2(args) {
  await consumeWithReconnect(
    'StreamL2Book',
    () => createClient().StreamL2Book(l2Request(args), createMetadata()),
    (update, count) => {
      const bestBid = update.bids && update.bids.length ? update.bids[0] : null;
      const bestAsk = update.asks && update.asks.length ? update.asks[0] : null;
      console.log(`[${count}] L2 ${update.coin} block=${update.block_number} bid=${levelText(bestBid)} ask=${levelText(bestAsk)} bids=${update.bids.length} asks=${update.asks.length}`);
    },
    args.maxMessages
  );
}

async function streamL4(args) {
  let snapshotCount = 0;
  await consumeWithReconnect(
    'StreamL4Book',
    () => createClient().StreamL4Book({ coin: args.coin }, createMetadata()),
    (update, count) => {
      if (update.snapshot) {
        snapshotCount += 1;
        const reset = l4SnapshotResetKind(snapshotCount);
        console.log(`[${count}] L4 snapshot ${update.snapshot.coin} height=${update.snapshot.height} reset=${reset} bids=${update.snapshot.bids.length} asks=${update.snapshot.asks.length}`);
        if (reset === 'replacement') {
          console.log('  replace the entire local L4 book with this snapshot');
        }
      } else if (update.diff) {
        let data;
        try {
          data = JSON.parse(update.diff.data || '{}');
        } catch (err) {
          console.warn(`[${count}] L4 diff height=${update.diff.height} invalid JSON: ${err.message}`);
          return;
        }
        console.log(`[${count}] L4 diff height=${update.diff.height} order_statuses=${(data.order_statuses || []).length} book_diffs=${(data.book_diffs || []).length}`);
      }
    },
    args.maxMessages
  );
}

async function streamBbo(args) {
  await consumeWithReconnect(
    'StreamBboBook',
    () => createClient().StreamBboBook({ coins: args.coins }, createMetadata()),
    (update, count) => {
      console.log(`[${count}] BBO ${update.coin} block=${update.block_number} bid=${levelText(update.bid)} ask=${levelText(update.ask)}`);
    },
    args.maxMessages
  );
}

async function streamL2Diff(args) {
  await consumeWithReconnect(
    'StreamL2BookDiff',
    () => createClient().StreamL2BookDiff(l2DiffRequest(args), createMetadata()),
    (update, count) => {
      console.log(`[${count}] L2 diff height=${update.height} snapshot=${update.snapshot} coins=${update.diffs.length}`);
      update.diffs.slice(0, 5).forEach(diff => {
        console.log(`  ${diff.coin} seq=${diff.seq} prev_seq=${diff.prev_seq} snapshot=${diff.snapshot} bid_changes=${diff.bids.length} ask_changes=${diff.asks.length}`);
      });
    },
    args.maxMessages
  );
}

async function streamL4Updates(args) {
  await consumeWithReconnect(
    'StreamL4BookUpdates',
    () => createClient().StreamL4BookUpdates({ coins: args.coins }, createMetadata()),
    (update, count) => {
      console.log(`[${count}] L4 updates height=${update.height} snapshot=${update.snapshot} diffs=${update.diffs.length}`);
      if (update.snapshot) console.log('  clear local L4 order state before applying this update');
      update.diffs.slice(0, 5).forEach(diff => {
        console.log(`  ${diff.diff_type} ${diff.coin} oid=${diff.oid} side=${diff.side || 'n/a'} px=${diff.px || 'n/a'} sz=${diff.sz || 'n/a'}`);
      });
    },
    args.maxMessages
  );
}

async function streamTpsl(args) {
  await consumeWithReconnect(
    'StreamTpslUpdates',
    () => createClient().StreamTpslUpdates({ coins: args.coins }, createMetadata()),
    (update, count) => {
      console.log(`[${count}] TP/SL height=${update.height} snapshot=${update.snapshot} diffs=${update.diffs.length}`);
      update.diffs.slice(0, 5).forEach(diff => {
        console.log(`  ${diff.diff_type} ${diff.coin} oid=${diff.oid} trigger=${diff.trigger_px || 'n/a'} limit=${diff.limit_px || 'n/a'} sz=${diff.sz || 'n/a'} reason=${diff.reason || 'n/a'}`);
      });
    },
    args.maxMessages
  );
}

async function main() {
  const args = parseArgs();
  validateArgs(args);

  console.log('Hyperliquid Orderbook Stream Example');
  console.log(`Endpoint: ${GRPC_ENDPOINT}`);
  console.log(`Mode: ${args.mode}`);
  const display = coinDisplay(args);
  console.log(`${display.label}: ${display.value}`);

  if (AUTH_TOKEN === 'your-quicknode-token') {
    console.error('Set AUTH_TOKEN to your QuickNode token before running this example.');
    process.exit(1);
  }

  switch (args.mode) {
    case 'l2':
      return streamL2(args);
    case 'l4':
      return streamL4(args);
    case 'bbo':
      return streamBbo(args);
    case 'l2-diff':
      return streamL2Diff(args);
    case 'l4-updates':
      return streamL4Updates(args);
    case 'tpsl':
      return streamTpsl(args);
    default:
      console.error('Invalid mode. Use --mode=l2, l4, bbo, l2-diff, l4-updates, or tpsl.');
      process.exit(1);
  }
}

if (require.main === module) {
  main().catch(err => {
    console.error('Stream failed:', err.message);
    process.exit(1);
  });
}

module.exports = { channelOptions, l4SnapshotResetKind };
